### ISSUE TITLE
S3 upload Key does not match AWS SDK

### ISSUE THREAD

**Issue by good-idea** [localstack/localstack#1180]
When using a localstack endpoint, S3 uploads will throw an error if you don't include the bucket name in the `Key`.

This works when uploading (via the node `aws-sdk`) to an actual S3 instance:

```js
s3client.upload(
	{
		Bucket: 'my-bucket',
		Key: `myfile.txt`,
		Body: data,
	},
```	

This uploads a file to `http://my-bucket.localhost:4572/myfile.txt`

To do this with Localstack, I need to do

```js
s3client.upload(
	{
		Bucket: 'my-bucket',
		Key: `my-bucket/myfile.txt`,
		Body: data,
	},
```

`http://my-bucket.localhost:4572/my-bucket/myfile.txt`

If I try this without the bucket name in the key, I get: `NoSuchBucket: The specified bucket does not exist`

localstack is running in a Docker container, and the logs don't tell me much:

`127.0.0.1 - - [09/Mar/2019 01:15:41] "PUT /myfile.txt HTTP/1.1" 404 -`

It's easy to 'fix' this by changing the Key when I am working locally. But, it would be nice if these matched up!

(Or, am I doing something wrong?)

**Comment by whummer** [comment]
Looks like this issue can be solved if you enable *path style addressing* on your S3 client - see this section in the README: https://github.com/localstack/localstack#troubleshooting

**Comment by good-idea** [comment]
Interesting, I didn't know that was an AWS setting. It looks like what Localstack is doing is still a little different than the default behavior - at least without any additional config for the client.

Instead of using either path-style (`localhost:4572/my-bucket/my-file.txt`) or virtual-hosted style, (`my-bucket.localhost:4572/my-file.txt`) it's doing both: requiring the bucket name as the start of the Key as well as using it for the subdomain.

In my case, it's an easy fix and I am good to go. It might still be confusing for others, though - it did take me a while to figure out what the problem was. I first tested my `s3client.upload` function on a new (live) AWS S3 bucket, and it worked - but when I switched to localstack I'd get the `NoSuchBucket` error.
